# 1. Data Collection

## 1.1 Import all Necessary Packages

In [14]:
import numpy as np

import matplotlib as mpl
import matplotlib.pyplot as plt

import pandas as pd

from datetime import date, time, datetime, timedelta 

## 1.2 Import the Raw Data

In [8]:
data = pd.read_csv("philadelphia_2019.csv")
weatherData = pd.read_csv("weather_hourly_philadelphia.csv")

## 1.3 Get an Overview over the Dataset

In [113]:
data.head()

start_time            end_time  start_station_id  end_station_id  \
0 2019-01-01 00:19:00 2019-01-01 00:27:00              3049            3007   
1 2019-01-01 00:30:00 2019-01-01 00:37:00              3005            3007   
2 2019-01-01 00:52:00 2019-01-01 01:05:00              3166            3169   
3 2019-01-01 00:55:00 2019-01-01 01:04:00              3058            3103   
4 2019-01-01 01:05:00 2019-01-01 01:17:00              3182            3028   

   bike_id user_type    start_station_name  \
0    14495  Indego30       Foglietta Plaza   
1     5332  Day Pass     Welcome Park, NPS   
2    14623  Indego30  Frankford & Belgrade   
3    11706  Indego30      20th & Fairmount   
4    11039  Indego30         17th & Sansom   

                            end_station_name        duration  start_weekday  \
0                     11th & Pine, Kahn Park 0 days 00:08:00              1   
1                     11th & Pine, Kahn Park 0 days 00:07:00              1   
2                                 2nd & Race 0 days 00:13:00              1   
3  27th & Master, Athletic Recreation Center 0 days 00:09:00              1   
4                           4th & Bainbridge 0 days 00:12:00              1   

   end_weekday  
0            1  
1            1  
2            1  
3            1  
4            1

In [104]:
data.describe()

start_station_id  end_station_id        bike_id  \
count     744260.000000   744260.000000  744260.000000   
mean        3081.936112     3081.409467    9825.729611   
std           58.702453       59.296591    5404.388722   
min         3000.000000     3000.000000       1.000000   
25%         3033.000000     3032.000000    3729.000000   
50%         3062.000000     3062.000000   11744.000000   
75%         3123.000000     3124.000000   14592.000000   
max         3210.000000     3210.000000   20265.000000   

                        duration  start_weekday    end_weekday  
count                     744260  744260.000000  744260.000000  
mean   0 days 00:24:44.364391476       2.841265       2.840554  
std    0 days 02:03:47.303420653       1.935174       1.936088  
min            -1 days +23:06:39       0.000000       0.000000  
25%              0 days 00:07:00       1.000000       1.000000  
50%              0 days 00:11:00       3.000000       3.000000  
75%              0 days 00:19:00       4.000000       4.000000  
max             21 days 13:24:08       6.000000       6.000000

In [12]:
# get all different user types

user_type_list = data['user_type'].unique()
user_type_list

array(['Indego30', 'Day Pass', 'Indego365', 'IndegoFlex', 'Walk-up', nan],
      dtype=object)

# 2. Data Preparation

## 2.1 Formatting and Adding Attributes

In [77]:
# add durations of each ride

data['duration'] = data['end_time'] - data['start_time']

In [78]:
# add start and end weekdays

def get_weekday (ts):
    return ts.weekday()

data['start_weekday']=data['start_time'].apply(lambda x: get_weekday (x))
data['end_weekday']=data['end_time'].apply(lambda x: get_weekday (x))

## 2.2 Dealing with Missing Values

In [66]:
# get number of missing values

data.isnull().sum()

start_time             0
end_time               0
start_station_id       0
end_station_id         0
bike_id                0
user_type             35
start_station_name     0
end_station_name       0
dtype: int64

because there are only 35 out of 744260 rows with missing values, we can simply drop these rows

In [68]:
# drop rows with missing values

data_clean = data.dropna(axis=0, inplace=False)

## 2.3 Dealing with Outliers

Possible outliers are rides with a very high or low duration

In [106]:
data.duration.describe()

count                       744260
mean     0 days 00:24:44.364391476
std      0 days 02:03:47.303420653
min              -1 days +23:06:39
25%                0 days 00:07:00
50%                0 days 00:11:00
75%                0 days 00:19:00
max               21 days 13:24:08
Name: duration, dtype: object

# 3. Descriptive Analytics

## 3.1 Temporal Demand Patterns and Seasonality

### 3.1.1 Average Usage per Weekday

### 3.1.2 Trip Duration per Weekday

In [127]:
# get average trip duration per weekday

duration_monday = data.duration[(data.start_weekday == 0)].sum() / data.duration[(data.start_weekday == 0)].count()
duration_tuesday = data.duration[(data.start_weekday == 1)].sum() / data.duration[(data.start_weekday == 1)].count()
duration_wednesday = data.duration[(data.start_weekday == 2)].sum() / data.duration[(data.start_weekday == 2)].count()
duration_thursday = data.duration[(data.start_weekday == 3)].sum() / data.duration[(data.start_weekday == 3)].count()
duration_friday = data.duration[(data.start_weekday == 4)].sum() / data.duration[(data.start_weekday == 4)].count()
duration_saturday = data.duration[(data.start_weekday == 5)].sum() / data.duration[(data.start_weekday == 5)].count()
duration_sunday = data.duration[(data.start_weekday == 6)].sum() / data.duration[(data.start_weekday == 6)].count()

In [ ]:
# plot



### 3.1.3 Seasonal Usage

### 3.1.4 Seasonal Patterns of Usage on Weekdays

### 3.1.5 Temperature and Usage Correlation

## 3.2 Geographical Demand Patterns

## 3.3 Key Performance Indicators

# 4. Predictive Analytics